include/deal.II/dofs/dof_accessor.templates.h

```diff
@@ -2875,71 +2173,64 @@ namespace internal
         Assert(static_cast<unsigned int>(accessor.level()) <
-                 accessor.dof_handler->levels.size(),
+                 accessor.dof_handler->hp_cell_future_fe_indices.size(),
                ExcMessage("DoFHandler not initialized"));
 
-        return accessor.dof_handler->levels[accessor.level()]->future_fe_index(
-          accessor.present_index);
+        if (future_fe_index_set(accessor))
+          return accessor.dof_handler
+            ->hp_cell_future_fe_indices[accessor.level()]
+                                       [accessor.present_index];
+        else
+          return accessor.dof_handler
+            ->hp_cell_active_fe_indices[accessor.level()]
+                                       [accessor.present_index];
       }
 
 
       /**
        * Do what the set_future_fe_index function in the parent class is
        * supposed to do.
        */
-      template <int dim, int spacedim, bool level_dof_access>
-      static void
-      set_future_fe_index(const DoFCellAccessor<DoFHandler<dim, spacedim>,
-                                                level_dof_access> &accessor,
-                          const unsigned int                       i)
-      {
-        (void)accessor;
-        (void)i;
-        // ::DoFHandler only supports a single active fe with index zero
-        Assert(
-          i == 0,
-          (typename std::decay<decltype(accessor)>::type::ExcInvalidObject()));
-      }
-
-
-
-      template <int dim, int spacedim, bool level_dof_access>
+      template <typename DoFHandlerType, bool level_dof_access>
       static void
       set_future_fe_index(
-        const DoFCellAccessor<dealii::hp::DoFHandler<dim, spacedim>,
-                              level_dof_access> &accessor,
-        const unsigned int                       i)
+        const DoFCellAccessor<DoFHandlerType, level_dof_access> &accessor,
+        const unsigned int                                       i)
       {
+        if (accessor.dof_handler->is_hp_dof_handler == false)
+          {
+            // ::DoFHandler only supports a single active fe with index zero
+            Assert(i == 0,
+                   (typename std::decay<decltype(
+                      accessor)>::type::ExcInvalidObject()));
+            return;
+          }
+
         Assert(
           accessor.dof_handler != nullptr,
           (typename std::decay<decltype(accessor)>::type::ExcInvalidObject()));
         Assert(static_cast<unsigned int>(accessor.level()) <
-                 accessor.dof_handler->levels.size(),
+                 accessor.dof_handler->hp_cell_future_fe_indices.size(),
                ExcMessage("DoFHandler not initialized"));
 
-        accessor.dof_handler->levels[accessor.level()]->set_future_fe_index(
-          accessor.present_index, i);
+        // accessor.dof_handler->levels_hp[accessor.level()]->set_future_fe_index(
+        //  accessor.present_index, i);
```

Comment:
Is this still needed? There are also a few todos left further down in this file.